File: np_asm_sse_f32.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
import numpy as np

# ______________________________________________________________________________
# float operations with 1 or 2 input(s)
inst_list = [
    ("add_ps", 2, np.add),  # SSE
    ("add_ss", 2, None),
    ("div_ps", 2, np.divide),
    ("div_ss", 2, None),
    ("max_ps", 2, None),
    ("max_ss", 2, None),
    ("min_ps", 2, None),
    ("min_ss", 2, None),
    ("mul_ps", 2, np.multiply),
    ("mul_ss", 2, None),
    ("rcp_ps", 1, np.reciprocal),
    ("rcp_ss", 1, None),
    ("rsqrt_ps", 1, None),
    ("rsqrt_ss", 1, None),
    ("sqrt_ps", 1, np.sqrt),
    ("sqrt_ss", 1, None),
    ("sub_ps", 2, np.subtract),
    ("sub_ss", 2, None),
    ("cmpeq_ps", 2, lambda x, y: np.equal(x, y).astype(np.float32)),
    ("cmpeq_ss", 2, None),
    ("cmpge_ps", 2, lambda x, y: np.greater_equal(x, y).astype(np.float32)),
    ("cmpge_ss", 2, None),
    ("cmpgt_ps", 2, lambda x, y: np.greater(x, y).astype(np.float32)),
    ("cmpgt_ss", 2, None),
    ("cmple_ps", 2, lambda x, y: np.less_equal(x, y).astype(np.float32)),
    ("cmple_ss", 2, None),
    ("cmplt_ps", 2, lambda x, y: np.less(x, y).astype(np.float32)),
    ("cmplt_ss", 2, None),
    ("cmpord_ps", 2, None),
    ("cmpord_ss", 2, None),
    ("cmpunord_ps", 2, None),
    ("cmpunord_ss", 2, None),
    ("cmpneq_ps", 2, lambda x, y: np.not_equal(x, y).astype(np.float32)),
    ("cmpneq_ss", 2, None),
    ("cmpnge_ps", 2, None),
    ("cmpnge_ss", 2, None),
    ("cmpngt_ps", 2, None),
    ("cmpngt_ss", 2, None),
    ("cmpnle_ps", 2, None),
    ("cmpnle_ss", 2, None),
    ("cmpnlt_ps", 2, None),
    ("cmpnlt_ss", 2, None),
    ("andnot_ps", 2, None),
    ("and_ps", 2, None),
    ("or_ps", 2, None),
    ("xor_ps", 2, None),
    ("movehl_ps", 2, None),
    ("movelh_ps", 2, None),
    ("move_ss", 2, None),
    ("unpackhi_ps", 2, None),
    ("unpacklo_ps", 2, None),
    ("addsub_ps", 2, None),  # SSE3
    ("hadd_ps", 2, None),
    ("hsub_ps", 2, None),
    ("movehdup_ps", 1, None),
    ("moveldup_ps", 1, None),
]


def gen_c(file_name="np_asm_sse_f32_auto.c"):
    n_inst = len(inst_list)

    fid = open(file_name, "w")

    fid.write("""// FILE AUTO-GENERATED FROM PYTHON CODE - DO NOT EDIT!
// instf_str is the list of instruction names
#define N_INSTF %d

static const char *instf_str[N_INSTF] =
{""" % n_inst)
    fid.writelines(['''
    "_mm_%s", ''' % k[0] for k in inst_list])
    fid.write("""
};""")

    fid.write("""

// n_in_f is the number of inputs for each instruction
static const char n_in_f[N_INSTF] =
{""")
    fid.writelines(["""
    %d,""" % k[1] for k in inst_list])
    fid.write("""
};""")

    fid.write("""

// create a function np_<op> for each instruction <op>""")

    # note: use _mm_store_ps and _mm_load_ps
    # use _mm_storeu_ps and _mm_loadu_ps if alignment is an issue
    for inst_name, n_input, _ in inst_list:
        if n_input == 1:
            fun_str = """
static void np_%s(
    char **args, const npy_intp *dimensions, const npy_intp *steps, void *data)
{
    npy_intp n = dimensions[0];
    float *in1 = (float *)args[0];
    float *out = (float *)args[1];
    int size_ratio = sizeof(__m128)/sizeof(float);
    // steps[k] == sizeof(float)
    npy_intp i = n/size_ratio;
    while (i > 0)
    {
        i--;
        // BEGIN main ufunc computation
        _mm_store_ps(out, _mm_%s(_mm_load_ps(in1)));
        // END main ufunc computation
        in1 += size_ratio;
        out += size_ratio;
    }
}
""" % (inst_name, inst_name)
        else:
            fun_str = """
static void np_%s(
    char **args, const npy_intp *dimensions, const npy_intp *steps, void *data)
{
    npy_intp n = dimensions[0];
    float *in1 = (float *)args[0];
    float *in2 = (float *)args[1];
    float *out = (float *)args[2];
    int size_ratio = sizeof(__m128)/sizeof(float);
    // steps[k] == sizeof(float)
    npy_intp i = n/size_ratio;
    while (i > 0)
    {
        i--;
        // BEGIN main ufunc computation
        _mm_store_ps(out, _mm_%s(_mm_load_ps(in1), _mm_load_ps(in2)));
        // END main ufunc computation
        in1 += size_ratio;
        in2 += size_ratio;
        out += size_ratio;
    }
}
""" % (inst_name, inst_name)
        fid.write(fun_str)

    fid.write("""
// funf is the list of all npy_<op> functions
PyUFuncGenericFunction funf[N_INSTF][1] =
{""")
    fid.writelines(["""
    {&np_%s},""" % k[0] for k in inst_list])
    fid.write("""
};

static char typef[3] = {NPY_FLOAT, NPY_FLOAT, NPY_FLOAT};
""")
    fid.close()


if __name__ == "__main__":
    gen_c()

File: np_asm_sse_f32.c
//______________________________________________________________________________
// float operation with 1 or 2 input(s)
#define INSTF_LIST \
    etype(add_ps, macro_if2) \
    etype(add_ss, macro_if2) \
    etype(div_ps, macro_if2) \
    etype(div_ss, macro_if2) \
    etype(max_ps, macro_if2) \
    etype(max_ss, macro_if2) \
    etype(min_ps, macro_if2) \
    etype(min_ss, macro_if2) \
    etype(mul_ps, macro_if2) \
    etype(mul_ss, macro_if2) \
    etype(rcp_ps, macro_if1) \
    etype(rcp_ss, macro_if1) \
    etype(rsqrt_ps, macro_if1) \
    etype(rsqrt_ss, macro_if1) \
    etype(sqrt_ps, macro_if1) \
    etype(sqrt_ss, macro_if1) \
    etype(sub_ps, macro_if2) \
    etype(sub_ss, macro_if2) \
    etype(cmpeq_ps, macro_if2) \
    etype(cmpeq_ss, macro_if2) \
    etype(cmpge_ps, macro_if2) \
    etype(cmpge_ss, macro_if2) \
    etype(cmpgt_ps, macro_if2) \
    etype(cmpgt_ss, macro_if2) \
    etype(cmple_ps, macro_if2) \
    etype(cmple_ss, macro_if2) \
    etype(cmplt_ps, macro_if2) \
    etype(cmplt_ss, macro_if2) \
    etype(cmpord_ps, macro_if2) \
    etype(cmpord_ss, macro_if2) \
    etype(cmpunord_ps, macro_if2) \
    etype(cmpunord_ss, macro_if2) \
    etype(cmpneq_ps, macro_if2) \
    etype(cmpneq_ss, macro_if2) \
    etype(cmpnge_ps, macro_if2) \
    etype(cmpnge_ss, macro_if2) \
    etype(cmpngt_ps, macro_if2) \
    etype(cmpngt_ss, macro_if2) \
    etype(cmpnle_ps, macro_if2) \
    etype(cmpnle_ss, macro_if2) \
    etype(cmpnlt_ps, macro_if2) \
    etype(cmpnlt_ss, macro_if2) \
    etype(andnot_ps, macro_if2) \
    etype(and_ps, macro_if2) \
    etype(or_ps, macro_if2) \
    etype(xor_ps, macro_if2) \
    etype(movehl_ps, macro_if2) \
    etype(movelh_ps, macro_if2) \
    etype(move_ss, macro_if2) \
    etype(unpackhi_ps, macro_if2) \
    etype(unpacklo_ps, macro_if2) \

#define N_INSTF (55-4)

// instf_str is the list of instruction names
#define etype(x, macro_cond) #x,
static const char *instf_str[N_INSTF] = {INSTF_LIST};
#undef etype

// n_in_f is the number of inputs for each instruction
#define macro_if1 1
#define macro_if2 2
#define etype(x, macro_cond) macro_cond,
static const char n_in_f[N_INSTF] = {INSTF_LIST};
#undef etype
#undef macro_if2
#undef macro_if1

// create a function np_<op> for each instruction <op>
#define INSTF(inst, macro_cond) \
static void np_ ## inst( \
    char **args, const npy_intp *dimensions, const npy_intp *steps, void *data) \
{ \
    npy_intp i; \
    npy_intp n = dimensions[0]; \
    float *in1 = (float *)args[0]; \
    float *in2 = (float *)args[1]; \
    float *out = (float *)args[2]; \
    int size_ratio = sizeof(__m128)/sizeof(float); \
    npy_intp in1_step = steps[0]/size_ratio; \
    npy_intp in2_step = steps[1]/size_ratio; \
    npy_intp out_step = steps[2]/size_ratio; \
    for (i = 0; i < n/size_ratio; i++) { \
        /* BEGIN main ufunc computation */ \
        _mm_storeu_ps(out, _mm_ ## inst( \
            macro_cond(_mm_loadu_ps(in1), _mm_loadu_ps(in2)))); \
        /* END main ufunc computation */ \
        in1 += in1_step*size_ratio; \
        in2 += in2_step*size_ratio; \
        out += out_step*size_ratio; \
    } \
}

#define macro_if1(x, y) x
#define macro_if2(x, y) x, y
#define etype(x, macro_cond) INSTF(x, macro_cond)
INSTF_LIST
#undef etype
#undef macro_if2
#undef macro_if1

// funf is the list of all npy_<op> functions
#define etype(x, macro_cond) {& np_ ## x},
PyUFuncGenericFunction funf[N_INSTF][1] = {INSTF_LIST};
#undef etype

static char typef[3] = {NPY_FLOAT, NPY_FLOAT, NPY_FLOAT};
